Phase 5: Update Documentation (#177)
Ekumen-OS/lambkin: README.md

LAMBKIN separates the orchestration machinery from the benchmark definition. The algorithm runs as an external process — LAMBKIN does not need to know what is inside it. Parameter sweeps, process lifecycle, I/O, and metric collection are all handled by the SDK, so your script stays focused on the benchmark logic.

> SLAM algorithms are rarely deterministic — particle filters resample randomly, sensor noise varies run to run, and timing jitter between processes can shift outcomes on its own. A single run's accuracy metric is one sample from a noisy distribution, not a reliable estimate of how an algorithm actually performs. LAMBKIN treats `num_iterations` as a first-class part of every benchmark definition, not an afterthought, so a variant is always evaluated as a population of runs you can summarize — with a mean, a standard deviation, an RMSE — instead of a single number taken on faith.

Reviewer: I think paragraph "SLAM algorithms are rarely deterministic..." can be summarized and added into the table below as a capability, instead of having as a paragraph here. For example:

**Iterations** | Run each variant across multiple iterations, giving results statistical significance instead of relying on a single noisy sample.

Author: Solved in 23cef14

> The same reasoning extends to where results land on disk. A benchmark only stays reproducible and reportable if its output follows a structure you didn't have to invent for that particular run: `results/var_<n>/iter_<n>/`, with metadata recorded alongside whatever artifacts your commands produced. That predictability is what lets `lambkin.data` walk any benchmark's results without being told their shape in advance, and what lets a notebook, a report generator, or a teammate's script reuse the same data months later without reverse-engineering a one-off layout.

Reviewer: Trim this paragraph — it's mixing philosophy with implementation detail it doesn't need. The literal `results/var_<n>/iter_<n>/` pattern already has a home with the actual tree diagram in the SDK README's "Expected Output" section, and naming `lambkin.data` here duplicates the Capabilities table's "Result access" row. This section should just make the argument — predictable structure enables later reuse — without committing to the specific shape or API that backs it:

```suggestion
A benchmark's output deserves the same discipline as its execution: a structured, predictable layout, with metadata recorded alongside whatever artifacts your commands produce. That predictability is what lets a notebook, a report generator, or a teammate's script revisit any benchmark's output months later without reverse-engineering its shape.
```

Author: Solved in 24b200e

| Feature | Description |
|---|---|
| **Parameter sweeps** | Declare combinations of algorithms, datasets, and parameters. LAMBKIN runs each combination as an independent iteration. |
| **Scoped inputs** | Resolve data dependencies once at the benchmark, variant, or iteration level — not on every single run. |
| **Process lifecycle** | Launch, supervise, and terminate external processes automatically across benchmark iterations, with cgroup-based cleanup. |
| **Partial restarts** | Skip iterations that already completed successfully in a previous run, based on a hash of their inputs. |
| **Dry-run validation** | Validate an entire benchmark's flow — variants, inputs, process calls — without executing a single command, via `--dry-run`. |
| **Context passing** | Carry configuration, paths, and state through the pipeline without coupling stages to each other. |
| **Result access** | Read structured benchmark outputs (`lambkin.data`) from a notebook or a standalone script, with no live benchmark required. |

To understand how LAMBKIN works under the hood, see the [SDK documentation](src/lambkin/README.md).

## Prerequisites

- Python 3.10+
- [`uv`](https://github.com/astral-sh/uv)

> - Linux with cgroups v2 and systemd (required for background process management)

Reviewer: I do not think we require systemd (that was before when we were running processes via systemd run ...). Linux with cgroups v2 is sufficient.

Author: Solved in 0e1efb7

> ## Installation

Reviewer: Suggestion for the Installation after carefully thinking about this:

---

```bash
git clone -b next-gen git@github.com:Ekumen-OS/lambkin.git
cd lambkin
uv tool install .
```

This installs lambkin on your `PATH` — you can delete the cloned directory afterward, and rerun this command to pick up updates. Use `--editable` instead if you want it to track your local checkout live (then keep the clone in place).

Alternatively, you can install it into a uv-managed virtual environment:

```bash
uv sync
```

Then run benchmarks via `uv run lambkin my_benchmark.py`.

---

Author: Solved in a74d6d1

Reviewer: Missing the closing line "Then run benchmarks via `uv run lambkin my_benchmark.py`." I suggested to add. It is relevant, so please do not forget to add it @teresa-ortega

Author: Solved in 31b728f

```bash
git clone -b next-gen git@github.com:Ekumen-OS/lambkin.git
uv sync
```

## Usage

A minimal example composing LAMBKIN's primitives into a working benchmark.

```python
import lambkin

@lambkin.benchmark(
    variants=lambkin.common.named_product(
        param_a=["x", "y"],
        param_b=[1, 10, 100],
    ),
    num_iterations=5,
)
@lambkin.option("--clock-rate", default=1.0)
def my_benchmark(ctx):
    with lambkin.process.background(ctx.shell.my_recorder, "-o", "output.mcap"):
        with lambkin.process.background(
            ctx.shell.my_algorithm,
            f"param_a:={ctx.variant.param_a}",
            f"param_b:={ctx.variant.param_b}",
            f"input:={ctx.inputs.dataset}",
        ):
            ctx.shell.my_player(ctx.inputs.dataset, "-r", ctx.options.clock_rate)

    ctx.shell.evo_ape.bag2(
        "output",
        "/ground_truth",
        "/pose",
        "--save_results",
        "output.ape.zip",
    )

@my_benchmark.input
def dataset(ctx):
    return ctx.source.path.parent / "datasets" / "my_dataset.mcap"

> @my_benchmark.ouput

Reviewer: ```suggestion
@my_benchmark.output
```

Author: Solved in 0e1efb7

def plot(ctx):
    data = lambkin.data.evo.series(ctx)

if __name__ == "__main__":
    my_benchmark()
```

Run it with the CLI:

```bash
lambkin my_benchmark.py
```

> Run `lambkin my_benchmark.py --help` to see all available options.

Reviewer: I think is better without using the script here. Just general sdk usage. The help will already redirect towards how to derive the options registered in the script.
```suggestion
Run `lambkin --help` to see all available options.
```

Author: Solved 0e1efb7

For a complete, working example using the Beluga algorithm, see [`examples/beluga/beluga_benchmark.py`](examples/beluga/beluga_benchmark.py). For the full CLI reference, see the [SDK documentation](src/lambkin/README.md#cli).

> ## Use Cases

Reviewer: Missing more use cases like benchmarking different algorithms

Author: Solved in 1a49bb7

### Benchmarking different datasets

When the input itself depends on the variant — e.g. a different sensor model needs a different recording — scoping the hook to `"variant"` means it's only resolved when the variant changes, not on every iteration. This keeps a multi-dataset sweep just as cheap as a single-dataset one.

```python
@nominal.input(scope="variant")
def dataset(ctx):

>     return ctx.source.path.parent / "datasets" / f"{ctx.variant.sensor_model}.mcap"

Reviewer: ```suggestion
@lambkin.benchmark(
    variants=lambkin.common.named_product(
        dataset=["warehouse", "office"]
    ),
    num_iterations=2,
)
def nominal(ctx):
    ...

@nominal.input(scope="variant")
def dataset(ctx):
    return ctx.source.path.parent / "datasets" / f"{ctx.variant.dataset}.mcap"
```

Author: Solved in 0e1efb7

```

> ## Cookbook

Reviewer: Move this out of Results — it doesn't belong there. The snippet uses ctx.shell, which only exists inside the benchmark function during execution (output hooks get a BenchmarkContext with no shell, see the existing warning above). So this isn't a "read back finished results" recipe, it's "convert what this iteration's process just produced into a format evo can score" — that's the Execution moment, not Result access.

Given that, I'd make this a standalone top-level Cookbook section near the end of the SDK README (sibling to Results, not nested under it).

Also: frame it as wrapping evo's own multi-format support rather than a lambkin-specific converter. Maybe mention all format conversions supported and link the evo docu: https://github.com/MichaelGrupp/evo/wiki/Formats#saving--exporting-to-other-formats

Author: Solved in 1a49bb7

| **Iterations** | Run each variant across multiple iterations, giving results statistical significance instead of relying on a single noisy sample. |
| **Scoped inputs** | Resolve data dependencies once at the benchmark, variant, or iteration level — not on every single run. |
| **Process lifecycle** | Launch, supervise, and terminate external processes automatically across benchmark iterations, with cgroup-based cleanup. |
| **Partial restarts** | Skip iterations that already completed successfully in a previous run, based on a hash of their inputs. |
| **Dry-run validation** | Validate an entire benchmark's flow — variants, inputs, process calls — without executing a single command, via `--dry-run`. |
| **Context passing** | Carry configuration, paths, and state through the pipeline without coupling stages to each other. |
| **Result access** | Read structured benchmark outputs (`lambkin.data`) from a notebook or a standalone script, with no live benchmark required. |

To understand how LAMBKIN works under the hood, see the [SDK documentation](src/lambkin/README.md).

## Prerequisites

- Python 3.10+
- [`uv`](https://github.com/astral-sh/uv)

> - Linux with cgroups v2 (required for background process management)

Reviewer: Not only background process management. We also use it for the script, foreground processes, etc. We can just say "Linux with cgroups v2" without adding the "(required for background process management)". Maybe link to cgroups v2 docu if you haven't done that already anywhere else: https://docs.kernel.org/admin-guide/cgroup-v2.html

Author: Solved in 40e7598

)
@lambkin.option("--clock-rate", default=1.0)
def my_benchmark(ctx):
    with lambkin.process.background(ctx.shell.my_recorder, "-o", "output.mcap"):
        with lambkin.process.background(
            ctx.shell.my_algorithm,
            f"param_a:={ctx.variant.param_a}",
            f"param_b:={ctx.variant.param_b}",
            f"input:={ctx.inputs.dataset}",
        ):
            ctx.shell.my_player(ctx.inputs.dataset, "-r", ctx.options.clock_rate)

    ctx.shell.evo_ape.bag2(
        "output",
        "/ground_truth",
        "/pose",
        "--save_results",
        "output.ape.zip",
    )

@my_benchmark.input
def dataset(ctx):
    return ctx.source.path.parent / "datasets" / "my_dataset.mcap"

@my_benchmark.output
def plot(ctx):
    data = lambkin.data.evo.series(ctx)

if __name__ == "__main__":
    my_benchmark()
```

Run it with the CLI:

```bash
lambkin my_benchmark.py
```
Run `lambkin --help` to see all available options.

For a complete, working example using the Beluga algorithm, see [`examples/beluga/beluga_benchmark.py`](examples/beluga/beluga_benchmark.py). For the full CLI reference, see the [SDK documentation](src/lambkin/README.md#cli).

> ## Use Cases

Reviewer: We could add the most basic use case, which is parameter sweeps. Then briefly explained that this can be done for programatic tuning of the algorithm parameters, to optimize, to catch regressions, etc.

Author: Solved in 2005f8c

### Benchmarking different datasets

When the input itself depends on the variant — e.g. a different sensor model needs a different recording — scoping the hook to `"variant"` means it's only resolved when the variant changes, not on every iteration. This keeps a multi-dataset sweep just as cheap as a single-dataset one.

```python
@lambkin.benchmark(
    variants=lambkin.common.named_product(dataset=["warehouse", "office"]),
    num_iterations=2,
)
def nominal(ctx):
    ...

@nominal.input(scope="variant")
def dataset(ctx):
    return ctx.source.path.parent / "datasets" / f"{ctx.variant.dataset}.mcap"
```

### Benchmarking different algorithms

> 
> `named_product` doesn't care whether a parameter is a tuning knob or a completely different code path. Adding an `algorithm` dimension to the sweep is enough to compare two localizers under identical conditions — same dataset, same number of iterations, same evaluation pipeline — with no extra orchestration:
> ```python

Reviewer: Spacing (new line).

```suggestion
`named_product` doesn't care whether a parameter is a tuning knob or a completely different code path. Adding an `algorithm` dimension to the sweep is enough to compare two localizers under identical conditions — same dataset, same number of iterations, same evaluation pipeline — with no extra orchestration:

```python
```

Author: Solved in d0c8f57

@lambkin.benchmark(
    variants=lambkin.common.named_product(
        algorithm=["beluga", "nav2_amcl"],
        sensor_model=["likelihood", "beam"],
        num_particles=[1, 10, 100, 1000, 10000],
    ),
    num_iterations=30,
)
@lambkin.option("--clock-rate", default=100.0)

> @lambkin.option("--clock", default=True)

Reviewer: I do not think we need to register options here. This is just a simple code snippet to showcase how to benchmark differents algortihms. Options usage is well documented.

Author: SOlved in c9257f6

Alternatively, install into a `uv`-managed virtual environment:

```bash
uv sync
```

## Usage

A minimal example composing LAMBKIN's primitives into a working benchmark.

```python
import lambkin

@lambkin.benchmark(
    variants=lambkin.common.named_product(
        param_a=["x", "y"],
        param_b=[1, 10, 100],
    ),
    num_iterations=5,
)
@lambkin.option("--clock-rate", default=1.0)
def my_benchmark(ctx):
    with lambkin.process.background(ctx.shell.my_recorder, "-o", "output.mcap"):
        with lambkin.process.background(
            ctx.shell.my_algorithm,
            f"param_a:={ctx.variant.param_a}",
            f"param_b:={ctx.variant.param_b}",
            f"input:={ctx.inputs.dataset}",
        ):
            ctx.shell.my_player(ctx.inputs.dataset, "-r", ctx.options.clock_rate)

    ctx.shell.evo_ape.bag2(
        "output",
        "/ground_truth",
        "/pose",
        "--save_results",
        "output.ape.zip",
    )

@my_benchmark.input
def dataset(ctx):
    return ctx.source.path.parent / "datasets" / "my_dataset.mcap"

@my_benchmark.output
def plot(ctx):

>     data = lambkin.data.evo.series(ctx)

Reviewer: Here it says plot but we do not plot. Might we worth to add the plot command just to have a more representative example.

Author: Solved in 7207f2e

| **Iterations** | Run each variant across multiple iterations, giving results statistical significance instead of relying on a single noisy sample. |
| **Scoped inputs** | Resolve data dependencies once at the benchmark, variant, or iteration level — not on every single run. |
| **Process lifecycle** | Launch, supervise, and terminate external processes automatically across benchmark iterations, with cgroup-based cleanup. |
| **Partial restarts** | Skip iterations that already completed successfully in a previous run, based on a hash of their inputs. |
| **Dry-run validation** | Validate an entire benchmark's flow — variants, inputs, process calls — without executing a single command, via `--dry-run`. |
| **Context passing** | Carry configuration, paths, and state through the pipeline without coupling stages to each other. |
| **Result access** | Read structured benchmark outputs (`lambkin.data`) from a notebook or a standalone script, with no live benchmark required. |

To understand how LAMBKIN works under the hood, see the [SDK documentation](src/lambkin/README.md).

## Prerequisites

- Python 3.10+
- [`uv`](https://github.com/astral-sh/uv)

> - [Linux with cgroups v2](https://docs.kernel.org/admin-guide/cgroup-v2.html)

Reviewer: ```suggestion
- Linux with [cgroups v2](https://docs.kernel.org/admin-guide/cgroup-v2.html)
```

Author: Solved in 6917903

## Use Cases

### Parameter sweeps

> The most basic use case: sweep a single parameter across a range of values, running every value under identical conditions. This is the basis for programmatic tuning — searching for the setting that optimizes a metric — and for regression testing, where re-running a fixed sweep over time surfaces any change that degrades performance.

Reviewer: Could be a single one or a set of them. Do not scope it down to just one.

Author: Solved in 6917903

```python
@lambkin.benchmark(
    variants=lambkin.common.named_product(num_particles=[100, 500, 1000, 5000]),
    num_iterations=30,
)
def nominal(ctx):

>     # run the algorithm configured with ctx.variant.num_particles

Reviewer: Mabe a reference to the actual algorithm call, for completeness. Like:

ctx.shell.my_algorithm(f"num_particles:={ctx.variant.num_particles}")
or
ctx.shell.my_algorithm(num_particles=ctx.variant.num_particles)

Author: Solved in 6917903
